discovery_engine_core/providers/src/gnews/models.rs

```diff
@@ -17,73 +17,72 @@ use chrono::{DateTime, Utc};
 use reqwest::Url;
 use serde::{Deserialize, Serialize};
 
-use crate::utils::deserialize_null_default;
+use crate::{utils::deserialize_null_default, Market};
 
 /// A news article
 #[derive(Clone, Serialize, Deserialize, Debug)]
-pub struct Article {
+pub(super) struct Article {
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub title: String,
+    pub(super) title: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub description: String,
+    pub(super) description: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub content: String,
+    pub(super) content: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub url: String,
+    pub(super) url: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub image: String,
+    pub(super) image: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub source: Source,
+    pub(super) source: Source,
 
     #[serde(rename(deserialize = "publishedAt"), alias = "date_published")]
-    pub published_at: DateTime<Utc>,
+    pub(super) published_at: DateTime<Utc>,
 }
 
 #[derive(Clone, Serialize, Deserialize, Debug, Default)]
-pub struct Source {
+pub(super) struct Source {
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub name: String,
+    pub(super) name: String,
 
     #[serde(deserialize_with = "deserialize_null_default")]
-    pub url: String,
+    pub(super) url: String,
 }
 
 /// Query response from the Gnews API
 #[derive(Deserialize, Debug)]
-pub(crate) struct Response {
+pub(super) struct Response {
     #[serde(default)]
-    pub(crate) articles: Vec<Article>,
+    pub(super) articles: Vec<Article>,
 
     /// Total articles available
     #[allow(dead_code)]
     #[serde(rename(deserialize = "totalArticles"))]
-    pub(crate) total_articles: usize,
+    pub(super) total_articles: usize,
 }
 
-impl From<Article> for crate::Article {
-    fn from(source: Article) -> Self {
-        let source_domain = Url::parse(&source.url)
+impl Article {
+    pub(super) fn into_generic_article(self, market: Market, topic: String) -> crate::Article {
```

Comment:
We should avoid using generics where not necessary as it increases compilation time, sure each case only to a negligible degree but it sums up. Compilation already takes too long.

I a bit less "clean" but I could accept `&Market` /`&str` as this will never be called with owned values.

Reply:
for local development compilation time shouldn't be much of an issue, mainly because of caching and because of realtime IDE feedback via rustanalyzer etc. a full clean of the rust targets is quite rare i would say. for the ci compilation time is not as important (still nice to be low but not crucial) and we also use more performant machines for that.

overall i would argue for easier to use interfaces over small compile time benefits.

Reply:
I would argue it _is_ already a problem even with caching and everything, but sure I will change it.

Reply:
Actually I would even say `impl Into<String>` and similar  might be anti-patterns. It hides costs, can mess with trait inference, adds compiler time, adds code size, can lead to unexpected conversions/cost in some cases, and it's only benefit is saving some `clone` calls.